### ISSUE TITLE
PDB Universal needs additional mechanisms for determining segment offsets

### ISSUE THREAD

**Issue by attikov** [NationalSecurityAgency/ghidra#3993]
**Describe the bug**
The default universal PDB parser does not add ImageBase to the symbol offsets.
If I change the parser to MSDIA, the symbols are fine.
Tested on a Delphi x86 PE where ImageBase is 400000

**To Reproduce**
1. New Project
2. Add a PE with PDB having ImageBase > 0
3. Load PDB with universal parser
4. Symbol table offsets are wrong

**Expected behavior**
Symbol table offsets are fine like loading with MSDIA

**Screenshots**
-

**Attachments**
-

**Environment (please complete the following information):**
 - OS:  W10
 - Java Version: -
 - Ghidra Version 10.1.2
 - Ghidra Origin: ghidra-sre.org

**Additional context**
 -

**Comment by ghizard** [comment]
I don't believe that I've tested a Delphi x86 PE.  The associated PDBs might look different than what we've seen and this might be difficult to diagnose.

Does it appear that the symbols are all just missing an addition of 0x400000, meaning that they are otherwise spaced out properly?  Or do they all seem clumped together as though they are all at the same address, such as 0x000000?

We attempt to add the ImageBase in the PDB Universal calculations, but there is a possiblity of something going wrong.  These calculations include both adding the image base as well as apply OMap table look-ups.  But if the calculations go wrong, there is a chance that we are returning one of several special address values, indicating a bad calculation, and these might look like zero.

Would you be able to run PdbDeveloperDumpScript.java (any program can be loaded) against the PDB to check for the existence of certain sections? (If I choose blah.pdb for input, I usually choose blah.pdb.txt for output.)  The output can be quite large.

Isolate all lines from `DebugData---------------------------------------------------` to `End DebugData-----------------------------------------------`.

To start, I'm wondering what streams are found in the StreamList.  Most particularly looking for non-ffff values for streams 04 and 05.  Details of what we are "able" to parse from any of these debug streams are found after the StreamList.

Another approach is to edit the launch.properties file to remove the # in front of `#VMARGS=-Dpdb.logging=true` and then relaunch Ghidra and try to do the analysis.  This will output some of this information to the .ghidra/.ghidra_version/pdb.analyzer.log file.

DebugData.java and PdbAddressManager.java are the Ghidra source code files that I'd likely have to fix.  Note that very little is known about the formats we are parsing with DebugData and much is missing.  PdbAddressManager is where we are attempting to use this data.

If you have the ability to debug the Ghidra code, you'd want to focus first on PdbAddressManager `Address getAddress(AddressMsSymbol symbol)` method.

**Comment by attikov** [comment]
Wow, thank you for the detailed reply. I'm not sure what did I see the other day but I tried again and the symbol table is full of 0x400000's offsets. I ran PdbDeveloperDumpScript.java, no sections in the DebugData but the addresses seems to be ok. As I can't debug it, I zipped the debug output, the exe and the pdb http://pisil.de/ghidra/demoproj.7z . Let me know if you downloaded it or if you won't download it that I can delete it.

**Comment by ghizard** [comment]
I was able to get your data. I see what you are saying about nothing in the DebugData section.

I started debugging and see multiple possible issues... one is where an offset is not being added (when there is no DebugData), but the other relates to the missing DebugData.  I'm trying to search my notes and some non-squashed commits from some saved older branches for code where I had originally tried parsing section information from the Linker module or from another place before settling on DebugData.  I believe I need to reincorporate that solution for when there is no DebugData, but want to see if I can find/reconstruct what I had done in the past before looking to do it anew.

**Comment by ghizard** [comment]
Update...

I've started making changes to try to make use of section symbol in the linker module, such as the following:
`SECTION: [0004], RVA = 00a2b000, Length = 0006D09C, Align = 0000000C, Characteristics = 40000000, .bss`
Note, that this RVA already has the image base added to it.

This works great, expect that two of the sections don't have typical RVA values.
For instance, `SECTION: [0005], RVA = 00000000, Length = 00000048, Align = 0000000C, Characteristics = 40000000, .tls`

The RVA is shown as 0x00000000, but MSDIA says it is 0x00a99000.  I could get this value from the program memory segment, but MSDIA is able to get it without having the program.  I'd like to be able to do the same.

I've debugged the PDB Universal parsing and processing to make sure fields were properly processed.

I've searched the *.pdb file for byte patterns of `00 90 a9 00` and `00 90 69 00` and traced these file locations back to internal streams and locations where the data should be parsed.  It turned out that one of these areas is the C13Lines section, which I had not yet tackled.  I started working on this and have more to do, but have gotten to the point where I've determined that these byte patterns are off-cut from C13 data and are not representative of segment offsets.  I'd like to complete this C13 work, but it still needs more study (between understanding the MSFT code that is available and speculating about code that they didn't make available)... there are supposedly some C13Lines COFF SYMBOL RVA records too, but I've not seen them in your PDB.

At the moment, I have the work broken down into two facets: the C13 work and the segment address determination work.  Both are a bit open-ended.  During this study, I also found a different PDB issues whose fix has recently been merged.
